src/solvers/smt2/smt2_conv.cpp

```diff
@@ -4314,7 +4324,28 @@ void smt2_convt::set_to(const exprt &expr, bool value)
 
   out << "; set_to " << (value?"true":"false") << "\n"
       << "(assert ";
-
+  if(!value)
+  {
+    out << "(not ";
+  }
+  const auto found_literal = defined_expressions.find(expr);
+  if(!(found_literal == defined_expressions.end()))
+  {
+    // This is a converted expression, we can just assert the literal name
+    // since the expression is already defined
+    out << found_literal->second;
+  }
+  else
+  {
+    convert_expr(prepared_expr);
+  }
+  if(!value)
+  {
+    out << ")";
+  }
+  out << ")\n";
+  return;
+  // OLD BELOW
```

Comment:
If this fix stays in then we should remove the old code below.

Reply:
Removed. Why didn't you just remove this to begin with? :satisfied:

Reply:
I have refactored the code and removed this comment.